### PR TITLE
[Merged by Bors] - feat: bounding integrals by asymptotics, part 1

### PR DESCRIPTION
**Pull request by llllvvuu**
part 1: corollaries such as for `atTop ℝ` yet to come

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/MeasureTheory/Integral/IntegrableOn.lean [hunk 1]
```diff
@@ -490,6 +490,12 @@ theorem IntegrableAtFilter.inf_ae_iff {l : Filter α} :
 alias ⟨IntegrableAtFilter.of_inf_ae, _⟩ := IntegrableAtFilter.inf_ae_iff
 #align measure_theory.integrable_at_filter.of_inf_ae MeasureTheory.IntegrableAtFilter.of_inf_ae
 
+@[simp]
+theorem IntegrableAtFilter.top : IntegrableAtFilter f ⊤ μ ↔ Integrable f μ := by
```
Comment (urkud) [inline Mathlib/MeasureTheory/Integral/IntegrableOn.lean:494]:
```suggestion
theorem integrableAtFilter_top : IntegrableAtFilter f ⊤ μ ↔ Integrable f μ := by
```

File: Mathlib/MeasureTheory/Integral/Asymptotics.lean [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  use s ∩ t, inter_mem hsl htl, hfm.restrict
+  refine ht.mono_set (inter_subset_right s t) |>.const_mul C' |>.2.mono
+    <| (hae_restrict_iff C' s hsl t htl).mpr <| ae_of_all _ fun x hx => (hs x hx.1).trans ?_
+  rewrite [norm_mul, hC']
+  gcongr
+  apply le_max_left
+
+/-- `Asymptotics.IsBigO.integrableAtFilter` with measurability hypothesis on `l` instead of `g`. -/
+theorem _root_.Asymptotics.IsBigO.integrableAtFilter'
+    (hfm : AEStronglyMeasurable f μ) (hgm : ∀ s ∈ l, MeasurableSet s)
```
Comment (urkud) [inline Mathlib/MeasureTheory/Integral/Asymptotics.lean:43]: The assumption `hgm` is never satisfied for a nontrivial filter.

File: Mathlib/MeasureTheory/Integral/Asymptotics.lean [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+## Main results
+
+* `Asymptotics.IsBigO.integrableAtFilter`: Given measurable functions `f`, `g` with `f = O[l] g`,
+  `IntegrableAtFilter g l μ` implies `IntegrableAtFilter f l μ`.
+-/
+
+open Asymptotics MeasureTheory Set Filter
+
+variable {α E : Type*} [MeasurableSpace α] [NormedDivisionRing E] [NormedSpace ℝ E]
+ {f g : α → E} {a b : α} {μ : Measure α} {l : Filter α}
```
Comment (urkud) [inline Mathlib/MeasureTheory/Integral/Asymptotics.lean:22]: Why do you need `E` to be a normed real algebra, not a normed space? Also, `f` and `g` can have different codomain.

File: Mathlib/MeasureTheory/Integral/Asymptotics.lean [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+open Asymptotics MeasureTheory Set Filter
+
+variable {α E F : Type*} [MeasurableSpace α] [NormedAddCommGroup E] [NontriviallyNormedField F]
+  {f : α → E} {g : α → F} {a b : α} {μ : Measure α} {l : Filter α}
+
+/-- Given `f = O[l] g` on measurably generated `l`,
+`IntegrableAtFilter g l μ`, and `f` `AEStronglyMeasurable` at `l`, `IntegrableAtFilter f l μ`. -/
+theorem _root_.Asymptotics.IsBigO.integrableAtFilter [IsMeasurablyGenerated l]
+    (hfm : ∃ s ∈ l, AEStronglyMeasurable f (μ.restrict s))
```
Comment (urkud) [inline Mathlib/MeasureTheory/Integral/Asymptotics.lean:27]: This assumption is exactly `StronglyMeasurableAtFilter f l μ`.

File: Mathlib/MeasureTheory/Integral/Asymptotics.lean (hunk 3 above)
Comment (llllvvuu) [reply Mathlib/MeasureTheory/Integral/Asymptotics.lean]: Thanks; I've managed to reduce it to `[NormedAddCommGroup E] [NontriviallyNormedField F]`

Comment (urkud) [reply Mathlib/MeasureTheory/Integral/Asymptotics.lean]: Should work for an `NormedAddCommGroup F`, if you add `IntegrableAtFilter.norm` and use it here.

Comment (llllvvuu) [reply Mathlib/MeasureTheory/Integral/Asymptotics.lean]: Right, of course 🤦‍♂️ ; fixed, thanks!

File: Mathlib/MeasureTheory/Integral/Asymptotics.lean [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    _ ≤ ∫⁻ (x : α) in S, C' * ‖g x‖₊ ∂μ := by
+      refine lintegral_mono_ae <| (ae_restrict_iff' hs_meas).mpr <| ae_of_all _ fun x hx => ?_
+      rewrite [← ENNReal.coe_mul, ENNReal.coe_le_coe]
+      refine (hs x (hs_le hx).1.1).trans ?_
+      show C * ‖g x‖ ≤ (max C 0) * ‖g x‖
+      gcongr
+      apply le_max_left
+    _ = C' * ∫⁻ (x : α) in S, ↑‖g x‖₊ ∂μ := lintegral_const_mul' _ _ ENNReal.coe_ne_top
+    _ < ⊤ := ENNReal.mul_lt_top ENNReal.coe_ne_top <| ne_top_of_lt
+      <| ht.mono_set (fun _ hx ↦ (hs_le hx).1.2) |>.2
```
Comment (urkud) [inline Mathlib/MeasureTheory/Integral/Asymptotics.lean:47]: You can avoid shrinking sets etc by using `Filter.smallSets` and available lemmas:
```suggestion
  have : ∀ᶠ s in l.smallSets, (∀ x ∈ s, ‖f x‖ ≤ C * ‖g x‖) ∧
      AEStronglyMeasurable f (μ.restrict s) ∧ IntegrableOn g s μ :=
    hC.smallSets.and <| hfm.eventually.and hg.eventually
  rcases this.exists_measurable_mem_of_smallSets with ⟨s, hsl, hsm, hfg, hf, hg⟩
  refine ⟨s, hsl, (hg.norm.const_mul C).mono hf ?_⟩
  refine (ae_restrict_mem hsm).mono fun x hx ↦ ?_
  exact (hfg x hx).trans (le_abs_self _)
```

Comment (llllvvuu) [reply Mathlib/MeasureTheory/Integral/Asymptotics.lean]: beautiful!

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Pull request closed without merging**